src/npyiter.rs

```diff
@@ -0,0 +1,472 @@
+use crate::array::{PyArray, PyArrayDyn};
+use crate::npyffi::{
+    array::PY_ARRAY_API,
+    types::{NPY_CASTING, NPY_ORDER},
+    *,
+};
+use crate::types::TypeNum;
+use pyo3::{prelude::*, PyNativeType};
+
+use std::marker::PhantomData;
+use std::os::raw::*;
+use std::ptr;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum NpyIterFlag {
+    CIndex,
+    FIndex,
+    MultiIndex,
+    ExternalLoop,
+    CommonDtype,
+    RefsOk,
+    ZerosizeOk,
+    ReduceOk,
+    Ranged,
+    Buffered,
+    GrowInner,
+    DelayBufAlloc,
+    DontNegateStrides,
+    CopyIfOverlap,
+    ReadWrite,
+    ReadOnly,
+    WriteOnly,
+}
+
+impl NpyIterFlag {
+    fn to_c_enum(&self) -> npy_uint32 {
+        use NpyIterFlag::*;
+        match self {
+            CIndex => NPY_ITER_C_INDEX,
+            FIndex => NPY_ITER_C_INDEX,
+            MultiIndex => NPY_ITER_MULTI_INDEX,
+            ExternalLoop => NPY_ITER_EXTERNAL_LOOP,
+            CommonDtype => NPY_ITER_COMMON_DTYPE,
+            RefsOk => NPY_ITER_REFS_OK,
+            ZerosizeOk => NPY_ITER_ZEROSIZE_OK,
+            ReduceOk => NPY_ITER_REDUCE_OK,
+            Ranged => NPY_ITER_RANGED,
+            Buffered => NPY_ITER_BUFFERED,
+            GrowInner => NPY_ITER_GROWINNER,
+            DelayBufAlloc => NPY_ITER_DELAY_BUFALLOC,
+            DontNegateStrides => NPY_ITER_DONT_NEGATE_STRIDES,
+            CopyIfOverlap => NPY_ITER_COPY_IF_OVERLAP,
+            ReadWrite => NPY_ITER_READWRITE,
+            ReadOnly => NPY_ITER_READONLY,
+            WriteOnly => NPY_ITER_WRITEONLY,
+        }
+    }
+}
+
+pub struct NpyIterBuilder<'py, T> {
+    flags: npy_uint32,
+    array: &'py PyArrayDyn<T>,
+}
+
+impl<'py, T: TypeNum> NpyIterBuilder<'py, T> {
+    pub fn new<D: ndarray::Dimension>(array: &'py PyArray<T, D>) -> NpyIterBuilder<'py, T> {
+        NpyIterBuilder {
+            flags: 0,
```

Comment:
Since `flags: 0` raise an exception, how about setting the default flag?
Or it might also good to have `NpyIterBuilder::writable` and `NpyIterBuilder::readonly` instead of `new`.

Reply:
Sounds good to make it `writable` and `readonly` methods instead of `new`.